tools/xabuild/XABuild.cs

```diff
@@ -21,15 +22,26 @@ static int Main ()
 				//Create a custom xabuild.exe.config
 				var xml = CreateConfig (paths);
 
-				//Hold open the file while creating the symbolic links
-				using (var writer = OpenSysLinksFile (paths)) {
-					//Create link to .NETFramework and .NETPortable directory
-					foreach (var dir in Directory.GetDirectories (paths.SystemProfiles)) {
-						var name = Path.GetFileName (dir);
-						if (!SymbolicLink.Create (Path.Combine (paths.FrameworksDirectory, name), dir)) {
-							return 1;
+				//Symbolic links to .NETFramework and .NETPortable directory
+				var systemDirectories = Directory.EnumerateDirectories (paths.SystemProfiles)
+					.Where (d => Path.GetFileName (d) != "MonoAndroid") //NOTE: this happened on one of our VSTS build agents
+					.ToArray ();
+				var symbolicLinks = systemDirectories
+					.Select (d => Path.Combine (paths.FrameworksDirectory, Path.GetFileName (d)))
+					.ToArray ();
+
+				if (symbolicLinks.Any (d => !Directory.Exists (d))) {
```

Comment:
This doesn’t make sense to me.

Imagine two `xabuild.exe` processes started at the “same” time, both hitting the above condition at the “same” time, with at least one directory not existing, causing this `if` to be entered.

What I believe would happen is that they *both* would execute the following `for` loop, creating the symbolic links. The “first” one would “win”, creating the symbolic links and the `.__sys_links.txt` file.

The “second” one, I *suspect* (but have not tested), would later execute `OpenSysLinksFile()` — *clearing* the file contents! — then *exit* as`SymbolicLink.Create()` returns `false` (assuming that `CreateSymbolicLink()` fails if the target file already exists; the docs don’t specify).

The end state being that the 2nd `xabuild.exe` does *nothing*, *and* `.__sys_links.txt` is cleared.

This sounds like a bad combination. :-(

Reply:
Whoops, this needs to use `File.AppendText` instead of `File.CreateText`.

Reply:
Even with `File.AppendText()`, that just introduces duplication within `.__sys_links.txt`. Which might be fine, actually...

That doesn't address the 2nd `xabuild.exe` exiting once it fails to create a symlink because the symlink already exists.

Reply:
`File.AppendText` should throw an `IOException` if two processes try to call it at the same time:

![image](https://user-images.githubusercontent.com/840039/38585893-2e89c32c-3ce1-11e8-89e5-c39465d66a6c.png)

`SymbolicLink` has its own `Directory.Exists` check: https://github.com/xamarin/xamarin-android/blob/3f7eca469382a29e67ec799fce1e62510d2291f1/tools/xabuild/SymbolicLink.cs#L12

Do you think that is enough? I have been running this test after deleting the symlinks to test this: https://github.com/xamarin/xamarin-android/blob/0dc96b02d9392adc136dc92469b00df7cb999a41/src/Xamarin.Android.Build.Tasks/Tests/Xamarin.Android.Build.Tests/BuildTest.cs#L2422

Reply:
> Do you think [the `Directory.Exists()` check] is enough?

No. That's the fun of multithreading! There's no cross-process locking going on, so I can imagine the `Directory.Exists(source)` check returning false (it doesn't exist), then the process is suspended, the *second* process comes in and actually creates the symlink, and thus it exists before the first process creates the symlink, which fails.

...except we *do* have cross-process locking on Windows, because of the file access...

...except that *mono* on Linux & macOS does *not* support file sharing across process boundaries:


```csharp
using System;
using System.IO;

class App {
	public static void Main ()
	{
		var path = "foo.txt";
		var x = File.AppendText (path);
		Console.ReadLine ();
	}
}
```

Run the above app in two different shell windows within the same directory, and no exception is thrown.

Additionally, we *want* to use `xabuild.exe` for Linux & macOS builds, as part of moving away from xbuild...

TL;DR: I believe that instead of a `Directory.Exists()` check *before* the symlink creation, we should instead (additionally?) do the `Directory.Exists()` check when the symlink creation *fails*, and if `Directory.Exists()` is true, then "ignore" the symlink creation failure, because the symlink *does* exist.

Reply:
Ok I see now.

Maybe a small test project is worth it? `xabuild-Tests`? Assuming the entire test run only takes a second or something.

Reply:
I'm not opposed to unit tests, though I'm not entirely sure i see what they'd be testing/how you'd be testing it...